### PR TITLE
Runtime type

### PR DESCRIPTION
**Pull request by msimberg**
Note: this builds on #4606. Only the last commit belongs to this PR.

This adds a `runtime_type` enum that can be passed in the `init_params` struct, to choose which runtime to start. If the distributed runtime isn't enabled and it's requested an exception will be thrown during init. The default is to use the distributed runtime if available and otherwise the local one.

### CODE REVIEW

File: libs/init_runtime/include/hpx/hpx_init_params.hpp [hunk 1]
```diff
@@ -51,6 +51,17 @@ namespace hpx {
...
+    {
+        runtime_type_local,
+        runtime_type_distributed,
+#if defined(HPX_HAVE_DISTRIBUTED_RUNTIME)
+        runtime_type_best = distributed
+#else
+        runtime_type_best = local
+#endif
+    };
+
```
Comment (hkaiser) [inline libs/init_runtime/include/hpx/hpx_init_params.hpp:64]: For the other enums exposed through the API we have an additional API that can return the string representation of a given enum value. Should we add that here as well? This could be used for generating error messages.

File: libs/init_runtime/include/hpx/hpx_init_params.hpp [hunk 2]
```diff
@@ -51,6 +51,17 @@ namespace hpx {
     }    // namespace resource
     /// \endcond
 
+    enum runtime_type
```
Comment (hkaiser) [inline libs/init_runtime/include/hpx/hpx_init_params.hpp:54]: I would like to see our enums in APIs to be slowly migrated to `enum class` types. This would also homogenize names to something like `runtime_type::local`

File: libs/init_runtime/include/hpx/hpx_init_params.hpp [hunk 3]
```diff
@@ -51,6 +51,17 @@ namespace hpx {
...
+    {
+        runtime_type_local,
+        runtime_type_distributed,
+#if defined(HPX_HAVE_DISTRIBUTED_RUNTIME)
+        runtime_type_best = distributed
+#else
+        runtime_type_best = local
+#endif
```
Comment (hkaiser) [inline libs/init_runtime/include/hpx/hpx_init_params.hpp:62]: Should name this `..._default` instead?

Comment (msimberg) [reply libs/init_runtime/include/hpx/hpx_init_params.hpp:62]: Caught me. I had it as a scoped enum and used `best` since `default` won't work there. Then I changed it to an unscoped one to be consistent with the other enums, and so that I could use default. I missed this spot though. I'm currently lacking a good replacement for `default`.

File: libs/init_runtime/include/hpx/hpx_init_params.hpp (hunk 2 above)
Comment (msimberg) [reply libs/init_runtime/include/hpx/hpx_init_params.hpp:54]: Agreed.

File: libs/init_runtime/include/hpx/hpx_init_params.hpp (hunk 1 above)
Comment (msimberg) [reply libs/init_runtime/include/hpx/hpx_init_params.hpp:64]: That's a good idea. Will add that.

**Comment by msimberg** [comment]
> Previously we discussed to extend the runtime_mode enum instead of introducing yet another enum type. Would this make sense or is it preferrable to have those separated?

I don't remember the context, but if I made that statement it may have been because I didn't understand `runtime_mode` properly. I think the two enums are mostly orthogonal, except for the fact that with `runtime_type_local` the `runtime_mode` doesn't make a difference.

File: libs/init_runtime/include/hpx/hpx_init_params.hpp (hunk 3 above)
Comment (hkaiser) [reply libs/init_runtime/include/hpx/hpx_init_params.hpp:62]: `default_`?

Comment (hkaiser) [reply libs/init_runtime/include/hpx/hpx_init_params.hpp:62]: In any case, whatever we do here we should do for the other enums as well.

Comment (msimberg) [reply libs/init_runtime/include/hpx/hpx_init_params.hpp:62]: Absolutely. I don't like adding underscores just to avoid name conflicts, but it may be better than  any alternative name I'll come up with. I'll sleep on it and see if I can come up with something else but otherwise it can be `default_`.

Comment (hkaiser) [reply libs/init_runtime/include/hpx/hpx_init_params.hpp:62]: I wouldn't expect anybody using this value explicitly. This would be mostly for use as a default argument value on our API declarations.

**Comment by hkaiser** [comment]
> > Previously we discussed to extend the runtime_mode enum instead of introducing yet another enum type. Would this make sense or is it preferrable to have those separated?
> 
> I don't remember the context, but if I made that statement it may have been because I didn't understand runtime_mode properly. I think the two enums are mostly orthogonal, except for the fact that with runtime_type_local the runtime_mode doesn't make a difference.

Playing devils advocate here... Wouldn't that mean that adding `runtime_mode::local` to the `runtime_mode` enumeration is sufficient? The existing enum values simply imply distributed operation.

File: libs/init_runtime/include/hpx/hpx_init_params.hpp (hunk 3 above)
Comment (hkaiser) [reply libs/init_runtime/include/hpx/hpx_init_params.hpp:62]: For the record, we discussed this here: #4127

**Comment by hkaiser** [comment]
> Previously we discussed to extend the runtime_mode enum instead of introducing yet another enum type. Would this make sense or is it preferrable to have those separated?

For the record, we discussed this here: #4127

**Comment by msimberg** [comment]
My memory... thanks @hkaiser. You're absolutely right about extending `runtime_mode`. It's much simpler to do that instead.

**Comment by msimberg** [comment]
I just pushed some changes but have not tested them properly yet. I'll be back at this one...

**Comment by msimberg** [comment]
If we go with class enums, should we again gradually phase out the unscoped enums? Is there a good convention for doing this kind of compatibility? I can think of two ways:

```
enum class runtime_mode {
    console = 0,
    // ...
    #if compatibility
    runtime_mode_console = console,
    // ...
    #endif
}

#if compatibility
using enum runtime_mode;
#endif
```
but this is likely to introduce name conflicts.

The other would be:
```
enum class runtime_mode {
    console = 0,
    // ...
}

#if compatibility
static constexpr runtime_mode runtime_mode_console = runtime_mode::console;
// ...
#endif
```

I guess we wouldn't break too many people's code with the runtime mode since that's almost always default or specified on the command line. The partitioner mode is another one which probably has few users. If we're going all in with this there are definitely other enums that would break user code by becoming class enums.

**Comment by hkaiser** [comment]
> ```
> enum class runtime_mode {
>     console = 0,
>     // ...
> }
> 
> #if compatibility
> static constexpr runtime_mode runtime_mode_console = runtime_mode::console;
> // ...
> #endif
> ```

This is the way to go, I think.

File: libs/runtime_configuration/include/hpx/runtime_configuration/runtime_mode.hpp [hunk 4]
```diff
@@ -22,8 +22,8 @@ namespace hpx {
         runtime_mode_console = 0,    ///< The runtime is the console locality
         runtime_mode_worker = 1,     ///< The runtime is a worker locality
         runtime_mode_connect = 2,    ///< The runtime is a worker locality
-                                     ///< connecting late
```
Comment (hkaiser) [inline libs/runtime_configuration/include/hpx/runtime_configuration/runtime_mode.hpp:25]: This comment got lost.

File: wrap/src/hpx_wrap.cpp [hunk 5]
```diff
@@ -107,7 +107,6 @@ extern "C" int initialize_main(int argc, char** argv)
         hpx::init_params iparams;
         iparams.desc_cmdline = desc;
         iparams.cfg = cfg;
-        iparams.mode = hpx::runtime_mode_console;
```
Comment (hkaiser) [inline wrap/src/hpx_wrap.cpp:110]: What is the rationale of this change?

Comment (msimberg) [reply wrap/src/hpx_wrap.cpp:110]: To have the wrapping use the default runtime mode. `runtime_mode_console` tries to start the distributed runtime, but if it's disabled that will fail. I'd rather have it fall back to the local runtime just like `init` or `start` do. If the distributed runtime is enabled and the runtime mode is `runtime_mode_default` the command line handling will anyway update it to `runtime_mode_console`. There's probably a corner case I'm not aware of though. @NK-Nikunj would you happen to know why the runtime mode was explicitly set to `runtime_mode_console` here?

Comment (msimberg) [reply wrap/src/hpx_wrap.cpp:110]: To have the wrapping use the default runtime mode. `runtime_mode_console` tries to start the distributed runtime, but if it's disabled that will fail. I'd rather have it fall back to the local runtime just like `init` or `start` do. If the distributed runtime is enabled and the runtime mode is `runtime_mode_default` the command line handling will anyway update it to `runtime_mode_console`. There's probably a corner case I'm not aware of though. @NK-Nikunj would you happen to know why the runtime mode was explicitly set to `runtime_mode_console` here?

Comment (NK-Nikunj) [reply wrap/src/hpx_wrap.cpp:110]: @msimberg I explicitly set it to `runtime_mode_console` because @hkaiser asked me to take the parameters from the `init_globally` example. There was not much thought put into the config. But it may get tricky to have user defined runtime modes in the config rather than a single explicit mode.

Comment (msimberg) [reply wrap/src/hpx_wrap.cpp:110]: Thanks @NK-Nikunj! I don't think we need user defined configs in `hpx_wrap.cpp` in the sense that user code would set the runtime mode. `runtime_mode_default` will do the same without any changes from the user, but will let users pass `--hpx:connect` etc. on the command line to change the mode if desired.

**Comment by msimberg** [comment]
@hkaiser I changed `runtime_mode` to be an enum class, but I added a generic compatibility option for unscoped enum compatibility. I suggest we use that one for all enums that need to be changed, and remove the option in 2.0.0. I'll try to change some other enums before 1.5.0 as well.

**Comment by msimberg** [comment]
https://cdash.cscs.ch//testDetails.php?test=21992139&build=109075 seems to be unrelated (I've seen it on other PRs now as well). Seems like either the stacksize test or stacksize handling is still not quite fixed. I'm reopening #4648 as a reminder to look at it again.
